lms/djangoapps/completion/models.py

```diff
@@ -0,0 +1,72 @@
+"""
+Completion tracking and aggregation models.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import ugettext as _
+from model_utils.models import TimeStampedModel
+
+from openedx.core.djangoapps.xmodule_django.models import CourseKeyField, UsageKeyField
+
+
+def validate_ratio(value):
+    """
+    Verify that the passed value is between 0.0 and 1.0.
+    """
+    if not 0.0 <= value <= 1.0:
+        raise ValidationError(_('{value} must be between 0.0 and 1.0').format(value=value))
+
+
+class BlockCompletionManager(models.Manager):
+    """
+    Custom manager for BlockCompletion model.
+
+    Adds update_or_create method.
+    """
+
+    def update_or_create(self, user, course_key, block_key, completed):
+        """
+        Update the completed value for the specified record.
+        """
+
+        obj, isnew = self.get_or_create(
+            user=user,
+            course_key=course_key,
+            block_type=block_key.block_type,
+            block_key=block_key,
+            defaults={'completed': completed},
+        )
+        if not isnew and obj.completed != completed:
+            obj.completed = completed
+            obj.full_clean()
+            obj.save()
+        return obj
+
+
+class BlockCompletion(TimeStampedModel, models.Model):
+    """
+    Track completion of completable blocks.
+    """
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
+    block_type = models.CharField(max_length=64)
+    completed = models.FloatField(validators=[validate_ratio])
+
+    objects = BlockCompletionManager()
+
+    class Meta(object):
+        unique_together = [
+            ('user', 'course_key', 'block_type', 'block_key'),
```

Comment:
I think we should remove `block_type` from this `unique_together`, since it's derived from `block_key`, and if we ever tried to save two entries that differed only in their `block_type` value, we'd want that to raise an error since that's always an invalid case. With this constraint as written, such a situation would be accepted.

Reply:
In fact, perhaps only `user, block_key` should be the unique constraint, since the `course_key` is usually derived from the block_key as well. (Not sure exactly how that works with old/draft mongo though.)

Reply:
I think you're right that I should keep the unique index precise.  It will need to be (user, course_key, block_key) to handle old mongo.  The other indexes will not need to be unique (but could be, in some cases).

Reply:
Block type still needs to be included in the composite indexes.  Even though it's derived from block_type, it will be needed to create filtered separately, in order to create efficient aggregations that filter by block case (to meet edX's watched-videos need).`